### PR TITLE
Enable revised output directory hash suffix computation 

### PR DESCRIPTION
**Pull request by fmeum**
Flips `--experimental_output_directory_naming_scheme` and `--experimental_exec_configuration_distinguisher` to `diff_against_baseline` and `off`, respectively, which makes it so that Starlark transitions can return to previous configurations, thus improving cache hit rates.

Also changes the fixed platform suffix in the exec configuration from an empty string to `"exec"` for compatibility with logic that looks for the substring `-exec-` in paths to determine whether the current configuration is the exec configuration.

As a result of this commit, output directory paths can change in the following ways:
* `-ST-<hash>` suffixes can change or disappear and, in rare cases, appear where they previously didn't
* for exec configuration directories, `k8-opt-exec-2B5CBBC6` becomes `k8-opt-exec-ST-011b9bdc32ed`

This may result in tool paths that are seven characters longer, which can cause builds on Windows to fail that were already very close to the `MAX_PATH` limit. A follow-up commit will reduce the length of the hash by three characters, bringing that increase down to four.

Work towards https://github.com/bazelbuild/bazel/issues/14023

### CODE REVIEW

**Comment by fmeum** [comment]
@gregestren

**Comment by fmeum** [comment]
Stacked on #16916

**Comment by fmeum** [comment]
@meteorcloudy As explained in https://github.com/bazelbuild/bazel/issues/14023#issuecomment-1325397261, this change would be rather important for rules_go in order to support Go 1.20 (due February 2023). It also generally improves build times and reduces target graph sizes for transition-heavy builds.

This flag only affects the `-ST-<hash>` part of output paths, with the hash being an intentionally undocumented implementation detail. @gregestren arrived at the conclusion that this would be to safe to cherry-pick in a minor release in https://github.com/bazelbuild/bazel/issues/14023#issuecomment-1335726213. 

Would it be possible to get a downstream pipeline for this change? It is currently stacked on https://github.com/bazelbuild/bazel/pull/16916, which is a fix for an unrelated bug that caused Java tools to fail if their paths get too long (the flag flip can make paths slightly longer or shorter, but only in ways that could already happen today via changes to the top-level config).

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by meteorcloudy** [comment]
A downstream test on this change is initiated here: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2765

Note that there are already some failures with Bazel@HEAD: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2764 (we're looking into fixing them)

**Comment by keertk** [comment]
@bazel-io fork 6.1.0

**Comment by fmeum** [comment]
Downstream test failures on Windows such as [this one](https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2765#0184e251-a958-4a9a-bf70-fc1992d0f9f3/294-454) *may* be a result of paths becoming longer as there are now two hash fragment in the exec configuration. 

@sdtwigg Given that the platform is already part of the ST hash (please correct me if I'm wrong here!), should I also flip `--experimental_exec_configuration_distinguisher` to `off` to mitigate this?

**Comment by fmeum** [comment]
@meteorcloudy Could you run another downstream test for this PR? I have made some changes to improve backwards compatibility and also reverted the commit causing https://github.com/bazelbuild/bazel/issues/17118 on this branch.

**Comment by sdtwigg** [comment]
Sorry for the delay, my github credentials got really messy so it took longer than expected to log back in again.

Can we keep the hash as ST for now? I have an excessive amount of documentation that refers to it as an ST hash that I don't want to update (or, moreso, would have to track down). Reducing the length is probably fine.

At the very least, if we were going to change the prefix from ST, would want to actually pick something meaningful.

**Comment by sdtwigg** [comment]
re: experimental_exec_configuration_distinguisher. I need to think about this and will get back to you.

My initial thoughts are I think it would be fine? But, I need to recall all the mechanics of the platform matching and exec transition to make sure. (I think the platform matching has exec platform outside the configuration, which is tricky, BUT the exec transition then puts the platform as the exec platform for the exec configuration so then is fine.)

**Comment by fmeum** [comment]
@meteorcloudy I temporarily rebased this PR onto `release-6.1.0`, taking @sdtwigg's remarks into account. Could you run a downstream pipeline for it?

If the pipeline passes, I would rebase this onto master and cherry-pick into `release-6.1.0` after review and merge.

**Comment by kshyanashree** [comment]
> @meteorcloudy I temporarily rebased this PR onto `release-6.1.0`, taking @sdtwigg's remarks into account. Could you run a downstream pipeline for it?
> 
> If the pipeline passes, I would rebase this onto master and cherry-pick into `release-6.1.0` after review and merge.

Hi @fmeum! With this change I have triggered the  downstream build here: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2852

**Comment by meteorcloudy** [comment]
Thanks @kshyanashree !

@fmeum The downstream pipeline currently has many existing breakages, please compare the result to the latest dowstream build https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds and check if additional failures are caused by this change.

BTW, did we already submit the same change at Bazel@HEAD? Can you explain a bit why aren't we cherry picking a submitted change?

**Comment by fmeum** [comment]
@meteorcloudy I am planning to submit this change at HEAD, I just first wanted to figure out whether it is compatible with Bazel 6.1, which e.g. still uses the native implementation of the Java rules. 

Looking at the downstream pipeline, I see that some targets fail to build on Windows because their paths got longer. I pushed https://github.com/bazelbuild/bazel/commit/36e7e37924a54f9d7211b91e21dc3415567c057f, which I hope will help with this by flipping another related flag. If that doesn't work, I can work on making the hashes even shorter.

@meteorcloudy Could you start a new downstream run at that commit? Generally speaking, are (minor) increases in the lengths of exec paths considered backwards incompatible changes (assuming I can get them so minor that the downstream tests still pass)?

**Comment by meteorcloudy** [comment]
I see, here is the downstream run: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2860

**Comment by fmeum** [comment]
@gregestren @sdtwigg Downstream tests look good, this is ready for review. When it has been merged to master, we can discuss whether it's still considered cherry-pickable into 6.1.0 (due in 1-2 weeks).

https://github.com/bazelbuild/bazel/pull/17474 will soften the blow to Windows by limiting the path length increase to four characters. We could discuss there whether the hash could be made even shorter, but that may require dropping the `ST-` prefix.

**Comment by gregestren** [comment]
I've looked at this. But I'm unsure what to make of the `experimental_exec_configuration_distinguisher` part, particularly because of @sdtwigg 's comments:

> re: experimental_exec_configuration_distinguisher. I need to think about this and will get back to you.

> My initial thoughts are I think it would be fine? But, I need to recall all the mechanics of the platform matching and exec transition to make sure. (I think the platform matching has exec platform outside the configuration, which is tricky, BUT the exec transition then puts the platform as the exec platform for the exec configuration so then is fine.)

I don't fully follow especially the platform mapping logic integrated with this. @sdtwigg as the author of those flags is the best expert - we need his followup.

File: src/main/java/com/google/devtools/build/lib/analysis/config/ExecutionTransitionFactory.java [hunk 1]
```diff
@@ -185,6 +185,7 @@ private static BuildOptions transitionImpl(BuildOptionsView options, Label execu
           break;
         default:
           // else if OFF do nothing
```
Comment (sdtwigg) [inline src/main/java/com/google/devtools/build/lib/analysis/config/ExecutionTransitionFactory.java:187]: Can you update this comment to "else if OFF just mark that we are now in an exec transition"

The mark shouldn't be necessary for correctness; however, it is very convenient for many classes of user to be able to look at the shortname and see that it is an exec configuration.

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/config/ExecutionTransitionFactory.java]: Done.

I have seen rulesets check `ctx.bin_dir.path` for `-exec-`, so I would assume removing that entirely would negatively affect downstream tests. It's also useful for humans though, I agree.

**Comment by sdtwigg** [comment]
Regarding my previous comments about safety:
I looked at my previous notes and tests and this should be fine now. (I was originally concerned about the exec configuration but it is actually safe to leave it unmarked and rely on the ST-hash system. Also, regardless, this PR changes the off behavior to mark it, which is very good for readability reasons alone.)

My only other concern is that, setting the aside the breaking issues reported with the output directories getting longer (although, part of me thinks those scenarios were already excessively fragile to be pushed over by this slight change), will the user-visible changes to output directories be too disruptive to their workflows for being in a minor release?

There are two sides:
A. Output directory shortname has never been strictly a specified or guaranteed so shifts to it are not considered breaking. (e.g. the ST-hash shifting between releases should be fine as it can occur for a variety of reasons)
B. By Hyrum's Law, users end up effectively relying on these sorts of things so such shifts can be disruptive. (e.g. with this flag on the ST-hash will appear in shortnames like all post-exec configs where it was previously avsent)

File: src/test/java/com/google/devtools/build/lib/rules/objc/ObjcRuleTestCase.java [hunk 2]
```diff
@@ -176,6 +180,29 @@ private String configurationDir(
       CompilationMode compilationMode) {
     String minOsSegment = minOsVersion == null ? "" : "-min" + minOsVersion;
     String modeSegment = compilationModeFlag(compilationMode);
+
```
Comment (sdtwigg) [inline src/test/java/com/google/devtools/build/lib/rules/objc/ObjcRuleTestCase.java:183]: Can you explain the changes to this file here and below?

**Comment by fmeum** [comment]
@sdtwigg I have also thought about this more and now feel that we should *not* cherry-pick this change. We haven't heard from rules_go users that they have seen their build times regress dramatically with Go 1.20 and as time passes, Bazel 7.0 comes closer and with it this commit.

**Comment by sdtwigg** [comment]
If we do not add this to 6.X, I do think we should submit this to 7.0.

(I'm a bit lost as to whether that means this is fine as-is being merged into bazelbuild:master or if another branch is more appropriate.)

**Comment by fmeum** [comment]
> (I'm a bit lost as to whether that means this is fine as-is being merged into bazelbuild:master or if another branch is more appropriate.)

It's fine to merge to master, Bazel 6.X releases are handled via the `bazel-6.X.Y` branches.

**Comment by fmeum** [comment]
@sdtwigg Just to make sure this isn't lost: If you want this PR to be merged, I think that you need to the `awaiting-merge` label.

**Review by sdtwigg — APPROVED** [review]

**Comment by com6056** [comment]
> @sdtwigg I have also thought about this more and now feel that we should _not_ cherry-pick this change. We haven't heard from rules_go users that they have seen their build times regress dramatically with Go 1.20 and as time passes, Bazel 7.0 comes closer and with it this commit.

@sdtwigg this is impacting us pretty badly, the recompiling of `GoStdlib` every time is pretty painful for us, any chance we can reconsider getting this cherry-picked instead into a Bazel 6 release? 🙏

**Comment by gregestren** [comment]
@com6056 what team do you represent?

Is it possible to manually set these flags for your own builds? Did rules_go need any other updates to its own code in response to these changes?

**Comment by com6056** [comment]
> @com6056 what team do you represent?
> 
> Is it possible to manually set these flags for your own builds? Did rules_go need any other updates to its own code in response to these changes?

I'm at [Mux](https://www.mux.com/), and what flags would we have to set? Is it just these?
```
--experimental_output_directory_naming_scheme=diff_against_baseline
--experimental_exec_configuration_distinguisher=off
```

If so, ya we can totally just set those for now in our `.bazelrc` if it is that easy 🎉

**Comment by fmeum** [comment]
I would recommend setting only the first one for now. The second may break some rulesets that rely on specific output path patterns to detect the exec configuration, which is fixed by the current PR.

**Comment by fmeum** [comment]
@gregestren @sdtwigg Friendly ping, is this in the process of being imported? Just want to make sure it hasn't gotten lost.

**Comment by gregestren** [comment]
@sdtwigg is still handling the import process. I last saw a code update from him earlier today.

**Comment by com6056** [comment]
Any updates on this? And is there any chance we could get the fix for using `--experimental_exec_configuration_distinguisher=off` cherry-picked into Bazel 6? 🙏

**Comment by com6056** [comment]
> I would recommend setting only the first one for now. The second one may break some rulesets that rely on specific output path patterns to detect the exec configuration, which is fixed by the current PR.

Do you know which rulesets are known to break with this as it is currently implemented in Bazel 6.1.1?

**Comment by gregestren** [comment]
@sdtwigg is the import ready?

**Comment by sdtwigg** [comment]
I had approved the import internally but it looks like it breaks some tests
and so was not actually submitted. I was not re-added to the attention set
when this happened so I had just been assuming everything was proceeding
fine. Planning to look into those test breakages ASAP. I speculate it is
due to those tests hard-coding lookups to intermediate artifacts.

**Comment by fmeum** [comment]
@sdtwigg https://github.com/bazelbuild/bazel/pull/17474 may be required to prevent downstream breakages in Bazel CI on Windows due to path length increases. That one may end up changing paths again.

**Comment by gregestren** [comment]
@sdtwigg have been having some back and forth on the internal tests. I think he's pretty close to unblocking them.

One of them is a shell test for path stripping. It assumed `bazel-out/k8-fastbuild/bin/...` on an Android binary. But these changes add a new hash to Android split transition that impacts Android binaries. So that needed a small update.

@fmeum we should export that test to GitHub. The only reason it's Google-internal is because only the Google executor has API support for path stripping before you changes, as you know.

**Pull request closed without merging**